Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift

```diff
@@ -140,92 +123,130 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
+                    videoWriter.cancelWriting()
+                    result = .failure(videoWriter.error ?? SentryOnDemandReplayError.errorRenderingVideo )
+                    group.leave()
+                    return
                 }
+                usedFrames.append(frame)
             }
+            frameCount += 1
         }
+        group.wait()
+        from = frameCount
+        
+        return try result?.get()
     }
-    
-    private func filterFrames(beginning: Date, end: Date) -> VideoFrames {
-        var framesPaths = [String]()
         
-        var screens = [String]()
+    private func finishVideo(outputFileURL: URL, usedFrames: [SentryReplayFrame], videoHeight: Int, videoWidth: Int, videoWriter: AVAssetWriter) -> Result<SentryVideoInfo?, Error> {
+        let group = DispatchGroup()
+        var finishError: Error?
+        var result: SentryVideoInfo?
         
-        var start = dateProvider.date()
-        var actualEnd = start
-        workingQueue.dispatchSync({
-            for frame in self._frames {
-                if frame.time < beginning { continue } else if frame.time > end { break }
-                
-                if frame.time < start { start = frame.time }
-                
-                if let screenName = frame.screenName {
-                    screens.append(screenName)
+        group.enter()
+        videoWriter.inputs.forEach { $0.markAsFinished() }
+        videoWriter.finishWriting {
+            defer { group.leave() }
+            if videoWriter.status == .completed {
+                do {
+                    let fileAttributes = try FileManager.default.attributesOfItem(atPath: outputFileURL.path)
+                    guard let fileSize = fileAttributes[FileAttributeKey.size] as? Int else {
+                        finishError = SentryOnDemandReplayError.cantReadVideoSize
+                        return
+                    }
+                    guard let start = usedFrames.min(by: { $0.time < $1.time })?.time else { return }
+                    let duration = TimeInterval(usedFrames.count / self.frameRate)
+                    result = SentryVideoInfo(path: outputFileURL, height: Int(videoHeight), width: Int(videoWidth), duration: duration, frameCount: usedFrames.count, frameRate: self.frameRate, start: start, end: start.addingTimeInterval(duration), fileSize: fileSize, screens: usedFrames.compactMap({ $0.screenName }))
+                } catch {
+                    finishError = error
                 }
-                
-                actualEnd = frame.time
-                framesPaths.append(frame.imagePath)
             }
+        }
+        group.wait()
+        
+        if let finishError = finishError { return .failure(finishError) }
+        return .success(result)
+    }
+    
+    private func filterFrames(beginning: Date, end: Date) -> [SentryReplayFrame] {
+        var frames = [SentryReplayFrame]()
+        workingQueue.dispatchSync({
```

Comment:
`m`: Why do we need `dispatchSync` here? Please add the info as a comment.

Reply:
```suggestion
        //Using dispatch queue as sync mechanism since we need a queue already to generate the video.
        workingQueue.dispatchSync({
```